### ISSUE TITLE
Amending order that breaches margin is cancelled

### ISSUE THREAD

**Issue by MM0819** [vegaprotocol/vega#5890]
Amending an order that would breach margin rules results in the order being cancelled rather than the existing order remaining in place

**Comment by jeremyletang** [comment]
@barnabee @davidsiska-vega what should be the expected behaviour here?

**Comment by davidsiska-vega** [comment]
Both behaviours make sense to me.

**Comment by davidsiska-vega** [comment]
If you're breaking margin rules then you're on my naughty list - don't expect me being nice ;-).

**Comment by MM0819** [comment]
FWIW, if we cancel the order then the behaviour on Vega would be unique and different to other exchanges

**Comment by barnabee** [comment]
We should refuse to amend the order but the amendment transaction should not cause an otherwise OK order to be cancelled.

Can we also currently close your position as a result of an amend that you can't afford to do or just cancel orders?

**Comment by barnabee** [comment]
The general principle should be that if we can't process your trans sanction then the state of the system including your orders and positions should be unchanged by it.

**Comment by davidsiska-vega** [comment]
> FWIW, if we cancel the order then the behaviour on Vega would be unique and different to other exchanges

Sounds like the more sensible behaviour is to leave the order in place then. 

> The general principle should be that if we can't process your transaction then the state of the system including your orders and positions should be unchanged by it. 

And based on this too.

**Comment by jeremyletang** [comment]
I'm starting work on that. I'm happy with implementing it as required. Just to confirm, can we think of any spamming related probem with someone trying to continuously amend an order which would not be allowed to be amend if margin is breached?

**Comment by barnabee** [comment]
> I'm starting work on that. I'm happy with implementing it as required. Just to confirm, can we think of any spamming related problem with someone trying to continuously amend an order which would not be allowed to be amend if margin is breached?

Hmmm… would that not be the same spam problem as successful amends? I could keep amending and have it work correctly and also fill up the block space?

So I assume this would need to fall under PoW spam prevention like other order commands, would that be sufficient?

cc @Vegaklaus

**Comment by jeremyletang** [comment]
> > I'm starting work on that. I'm happy with implementing it as required. Just to confirm, can we think of any spamming related problem with someone trying to continuously amend an order which would not be allowed to be amend if margin is breached?
> 
> Hmmm… would that not mean the same spam problem could happen with successful amends? I could keep amending and have it work correctly and also fill up the block space?
> 
> So I assume this would need to fall under PoW spam prevention like other order commands, would that be sufficient?
> 
> cc @Vegaklaus

that's fair. my point is maily here that many amend which are valid in the sense they don't breach margin, seems like an OK use case. But here it's easy to try to amend your orders with a delta of a int64 max size which will always fail margin check I would imagine and is clearly malicious.

**Comment by Vegaklaus** [comment]
For now, this can be handled with the normal Spam (same level as entering a huge amount of near-zero transactions). We should keep a not on that for antispam 2 though, where order volumes & overall generated revenue  will be used to gauge the likelihood of malice

**Comment by davidsiska-vega** [comment]
Should we stop processing the batch if the margin check fails on any of the components?

**Comment by barnabee** [comment]
> that's fair. my point is maily here that many amend which are valid in the sense they don't breach margin, seems like an OK use case. But here it's easy to try to amend your orders with a delta of a int64 max size which will always fail margin check I would imagine and is clearly malicious.

Yeah the question I suppose is how we tell where the line is between clearly malicious and not. TBH if we could reject the entire amend txn if it would definitely breach margin that would be fine, as it's only the fact that we cancel the order that is an issue. So I guess the change to not cancel the order is not directly related to the potential spam/malicious txn issue?

**Comment by jeremyletang** [comment]
> > that's fair. my point is maily here that many amend which are valid in the sense they don't breach margin, seems like an OK use case. But here it's easy to try to amend your orders with a delta of a int64 max size which will always fail margin check I would imagine and is clearly malicious.
> 
> Yeah the question I suppose is how we tell where the line is between clearly malicious and not. TBH if we could reject the entire amend txn if it would definitely breach margin that would be fine, as it's only the fact that we cancel the order that is an issue. So I guess the change to not cancel the order is not directly related to the potential spam/malicious txn issue?

yea it's not exactly related. BTW I've been looking at this earlier today, and wrote some tests for it. We actually just rejected on any margin breach, not even cancelling the existing order. So not sure where that came from haha. I can see only the case with pegged order, being totally cancelled though it margin was breached and the order was parked before. I suppose we should just rejected as well here and keep the order parked right?

**Comment by barnabee** [comment]
> yea it's not exactly related. BTW I've been looking at this earlier today, and wrote some tests for it. We actually just rejected on any margin breach, not even cancelling the existing order.

That seems like what we would want

> the case with pegged order, being totally cancelled though it margin was breached and the order was parked before. I suppose we should just rejected as well here and keep the order parked right?

Yeah that sounds right. May be worth getting @MM0819 to clarify what he was seeing exactly that was not good to make sure we are covering it with this.
